server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java

```diff
@@ -224,37 +247,97 @@ public synchronized void shutdownNode() {
+        // remove the node from the hub registry, if it reported 200 OK for NodeForceRestartServlet request
+        if (responseStatusCode == HttpStatus.SC_OK) {
+            getRegistry().removeIfPresent(this);
+        }
+        LOGGER.exiting();
     }
 
     /**
      * Thread will recycle the node when all active sessions are completed
      */
     class NodeRecycleThread extends Thread {
+        private volatile boolean running;
+
+        NodeRecycleThread() {
+            running = false;
+        }
 
         @Override
         public void run() {
-            appendMsgToCustomLog("Started NodeRecycleThread to restart the node once all the sessions are completed.");
-            while (getTotalUsed() > 0) {
+            LOGGER.entering();
+            running = true;
+            int timeout = -1; // Default. Wait forever.
+            int expired = 0;
+
+            try {
+                timeout = ConfigParser.parse().getInt("nodeRecycleThreadWaitTimeout");
+            } catch (ConfigParserException e) { // NOSONAR
+                // ignore we have a default already
+            }
+
+            proxyLogger.fine("Started NodeRecycleThread with " + ((timeout == -1) ? "no" : "a " + timeout + " second")
+                    + " timeout.");
+            while (keepLooping(expired, timeout)) {
                 try {
-                    Thread.sleep(1000);
+                    sleep(1000);
+                    expired += 1;
                 } catch (InterruptedException e) {
-                    LOGGER.log(Level.SEVERE, e.getMessage(), e);
+                    if (running) {
+                        //SEVERE, only if shutdown() was not called
+                        LOGGER.log(Level.SEVERE, e.getMessage(), e);
+                    }
+                    running = false;
+                    proxyLogger.warning("NodeRecycleThread was interrupted.");
+                    LOGGER.exiting();
+                    return;
                 }
             }
-            appendMsgToCustomLog("All sessions are completed and restarting the node.");
-            shutdownNode();
+
+            if (wasExpired(expired, timeout)) {
+                proxyLogger.info("Timeout occurred while waiting for sessions to complete. Shutting down the node.");
+            } else {
+                proxyLogger.info("All sessions are complete. Shutting down the node.");
+            }
+            forceNodeShutdown();
+            LOGGER.exiting();
+        }
+
+        private boolean keepLooping(int expired, int timeout) {
+            return (getTotalUsed() > 0) && (running) && ((expired < timeout) || (timeout == -1));
+        }
+
+        private boolean wasExpired(int expired, int timeout) {
+            return (expired >= timeout) && (timeout != -1);
+        }
+
+        public void shutdown() {
+            LOGGER.entering();
+            running = false;
+            proxyLogger.fine("Interrupting NodeRecycleThread.");
+            interrupt();
```

Comment:
We're setting running to false and interrupting the thread here.  Isn't one sufficient?